### PR TITLE
FEATURE: Reintroduce better thread reply counter cache

### PR DESCRIPTION
**Pull request by martin-brennan**
This was reverted in 38cebd3ed509524ad635adb107163d0496d0c550.
The issue was that I was using Discourse.redis.delete_prefixed
which does a slow redis KEYS lookup, which is not advised in
production. This commit removes that, and also ensures the periodical
thread count update only happens if threading is enabled.

I changed to use a redis INCR/DECR for reply count
cache. This avoids a round trip to redis to GET the current
count, and also avoids multi-process issues, where
if there's two processes trying to increment at the
same time, they may both receive the same value, add one
to it, then both write the same value back.
Then, it's only n+1 instead of n+2.

This also prevents almost all chat scheduled jobs from
running if chat is disabled, the only one remaining is
the message retention job.

### CODE REVIEW

File: plugins/chat/app/models/concerns/chat/thread_cache.rb [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+      def replies_count_cache_updated_at_redis_key(id)
+        "chat_thread:replies_count_cache_updated_at:#{id}"
+      end
+
+      def replies_count_cache_redis_key(id)
+        "chat_thread:replies_count_cache:#{id}"
+      end
+
+      def clear_caches!(ids)
+        ids = Array.wrap(ids)
```
Comment (martin-brennan) [inline plugins/chat/app/models/concerns/chat/thread_cache.rb:17]: Main change here is to not use `delete_prefixed` if there are no ids

File: plugins/chat/app/models/chat/thread.rb [hunk 2]
```diff
@@ -72,7 +79,10 @@ def self.update_counts
         ) subquery
         WHERE threads.id = subquery.thread_id
         AND subquery.replies_count != threads.replies_count
+        RETURNING threads.id AS thread_id;
       SQL
+      return if updated_thread_ids.empty?
```
Comment (martin-brennan) [inline plugins/chat/app/models/chat/thread.rb:84]: Don't need to bother trying to clear the cache if all the DB values are up to date

File: plugins/chat/app/jobs/regular/chat/update_thread_reply_count.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+module Jobs
+  module Chat
+    class UpdateThreadReplyCount < Jobs::Base
+      def execute(args = {})
+        return if !SiteSetting.enable_experimental_chat_threaded_discussions
```
Comment (martin-brennan) [inline plugins/chat/app/jobs/regular/chat/update_thread_reply_count.rb:7]: No need to even attempt to do this if threads aren't on

File: plugins/chat/app/models/chat/thread.rb [hunk 4]
```diff
@@ -50,6 +56,7 @@ def self.grouped_messages(thread_ids: nil, message_ids: nil, include_original_me
     end
 
     def self.ensure_consistency!
+      return if !SiteSetting.enable_experimental_chat_threaded_discussions
```
Comment (martin-brennan) [inline plugins/chat/app/models/chat/thread.rb:59]: No need to even attempt to do this if threads aren't on

**Review by danielwaterworth — APPROVED** [review]

**Pull request merged**